Implemented MSW for mocking api requests, axios for backend integration and React Query for data management

mainul-itransition opened the pull request:
**Overview**

This PR introduces Mock Service Worker (MSW) to mock API requests for local development, integrates Axios for backend requests, and utilizes React Query for data fetching, caching, and synchronization in our React TypeScript project.

**Changes Made**

1. **Setup Mock Service Worker (MSW)**

- Integrated MSW to mock API requests locally.
- Created handlers.ts to define mock endpoints (/api/dashboard).
- Implemented browser.ts to configure MSW worker setup.

2. **Backend Integration with Axios**

- Installed Axios for making HTTP requests.
- Created useDashboardData.ts hook to fetch data using Axios.
- Defined TypeScript types (DashboardData) for API responses.

3. **Data Fetching and Management with React Query**

- Installed @tanstack/react-query for managing data fetching, caching, and synchronization.
- Implemented useDashboardData custom hook using useQuery from React Query.

4. **UI Integration**

- Updated App.tsx to display fetched data using React Query.
- Added loading spinner from Ant Design (Spin) while data is fetching.

**Some Key Notes**

- I used latest React Query which is now Tanstack React Query.
- I tried to keep folders and things separate, added types, which we can adjust further.
- Currently kept the api function inside hooks folder , but I want to keep the api calling functions in a separate folder
- Currently didn't create axiosInstance as not needed for current use case.

Let me know if there are any additional changes or considerations needed.

mainul-itransition committed:
Mock Service worker Implementation for mocking api,fetched data using React Query and displayed data in app

alexeylos commented on src/App.tsx:
```diff
@@ -1,20 +1,36 @@
       </Flex>
+      <Card
+        title="Dashboard"
+        style={{ backgroundColor: '#ffffff', boxShadow: '0px 4px 8px rgba(0, 0, 0, 0.1)' }}
```
You can do that here as it's just a boilerplate but let's try to avoid inline styling in future. It's better to move all styles to `.less` files

mainul-itransition replied:
Okay. Got it.

alexeylos commented on src/types/types.ts:
I'd like to use another name for this file. `types/types` doesn't look good. Let's name typing files based on scope and something like `common.ts` for common types. This file can me named `dashboard.ts` so `types/dashboard.ts` contains types for Dashboard and it's easy to understand from its name

mainul-itransition replied:
Yes, you're right.This was just for getting insights from you, as I don't know what type of components or file will be or now.It's a great point.Thanks for your valuable insights.I'll make the changes accordingly.

mainul-itransition committed:
Implemented feedback: refactored types file naming, replaced inline styles in app.tsx with less

mainul-itransition commented:
Thank you for the feedback. I have made the following changes based on your suggestions:

1. Refactored the types folder file naming with scope to be more specific
2. Replaced inline styling in App.tsx file with less styling by writing css inside styles.less file
3. In future, we can create separate folder for less styles to keep scope based styling

Please review the changes and let me know if there are any further adjustments needed.

mainul-itransition committed:
tests added for app.tsx and merge conflicts resolved

mainul-itransition commented:
**Added Test cases for App.tsx and made changes in test coverage **

- Implemented test cases for App.tsx 
- changed test coverage to 20% 

**Key notes**

- faced type problem with jest mock mockReturnedValue

mainul-itransition committed:
fixed the jest config, package,json, refactored app.test.tsx,moved useDashboard to src/api/dashboard.ts

mainul-itransition commented:
**Updated changes in App.test.tsx, jest.config,package.json,src/api/dashboard.ts**

- updated collectCoverageFrom and test coverage threshold to 80
- updated yarn test  --findRelatedTests in package json
- created new folder api with dashboard.ts file inside which will handle the api calls and reuturn useDashboardData
- Removed queryclient provider etc from app.test.tsx
